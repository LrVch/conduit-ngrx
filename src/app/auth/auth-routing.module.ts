import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { AuthComponent } from './auth.component';
import { NoAuthGuard } from './no-auth-guard.service';

const routes: Routes = [
  {
    path: 'login',
    component: AuthComponent,
    canActivate: [NoAuthGuard],
    data: { title: 'conduit.menu.login', animation: 'AuthComponent' }
  },
  {
    path: 'register',
    component: AuthComponent,
    canActivate: [NoAuthGuard],
    data: { title: 'conduit.menu.register', animation: 'AuthComponent' }
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class AuthRoutingModule {}
