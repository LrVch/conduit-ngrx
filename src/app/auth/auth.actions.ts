import { Action } from '@ngrx/store';
import { Credentials } from '@app/core/models/credentials.model';
import { User, Errors } from '@app/core';

export enum AuthActionTypes {
  LoginPageAttemptLogin = '[Login Page] Attempt Login',
  LoginSuccess = '[Auth API] Login success',
  LoginFail = '[Auth API] Login fail',
  LoginAttemptCopmlete = '[Auth API] Login Attempt Copmlete',
  LoginPageSetAuthErrors = '[Auth API] Set Auth Errors',
  LoginPageClearAuthErrors = '[Login Page] Clear Auth Errors',
  LoggedLocalStorageRequest = '[LocalStorage] Login success Request',
  LoggedLocalStorage = '[LocalStorage] Login Success',
  ClearAuthErrors = '[Login Page] Clear Auth Errors',
  SettingsPageLogoutAction = '[Settings Page] Logout',
  LogoutConfirm = '[Auth] Logout Confirm',
  LogoutAction = '[Auth] Logout',
  AuthAttemptToGetUser = '[Auth] Try To Get User From Api',
  SetReturnUrl = '[Auth] Set Return Url',
  ClearReturnStateFromRouteChange = '[Auth] Clear Return State From Route Change',

  UpdateUserRequest = '[Auth API] Update User Request',
  UpdateUserSuccess = '[Auth API] Update User Success',
  UpdateUserFail = '[Auth API] Update User Fail',
  SetUpdateUserErrors = '[Auth API] Set Update User Errors'
}

export class SetReturnUrl implements Action {
  readonly type = AuthActionTypes.SetReturnUrl;

  constructor(public payload: { returnUrl: string }) { }
}

export class ClearReturnStateFromRouteChange implements Action {
  readonly type = AuthActionTypes.ClearReturnStateFromRouteChange;
}

export class AuthAttemptToGetUser implements Action {
  readonly type = AuthActionTypes.AuthAttemptToGetUser;
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.LoginSuccess;

  constructor(public payload: { user: User }) { }
}

export class LoggedLocalStorage implements Action {
  readonly type = AuthActionTypes.LoggedLocalStorage;

  constructor(public payload: { user: User }) { }
}

export class LoggedLocalStorageRequest implements Action {
  readonly type = AuthActionTypes.LoggedLocalStorageRequest;
}

export class LoginFail implements Action {
  readonly type = AuthActionTypes.LoginFail;

  constructor(public payload: { authErrors: Errors }) { }
}

export class LoginAttemptCopmlete implements Action {
  readonly type = AuthActionTypes.LoginAttemptCopmlete;
}

export class SettingsPageLogoutAction implements Action {
  readonly type = AuthActionTypes.SettingsPageLogoutAction;

  constructor(public payload: { question: string }) { }
}

export class LogoutConfirm implements Action {
  readonly type = AuthActionTypes.LogoutConfirm;
}

export class LogoutAction implements Action {
  readonly type = AuthActionTypes.LogoutAction;

  constructor(public payload?: { path?: string }) { }
}

export class LoginPageAttemptLogin implements Action {
  readonly type = AuthActionTypes.LoginPageAttemptLogin;

  constructor(public payload: { authType: string, credentials: Credentials }) { }
}

export class LoginPageClearAuthErrors implements Action {
  readonly type = AuthActionTypes.LoginPageClearAuthErrors;
}

export class ClearAuthErrors implements Action {
  readonly type = AuthActionTypes.ClearAuthErrors;
}


export class LoginPageSetAuthErrors implements Action {
  readonly type = AuthActionTypes.LoginPageSetAuthErrors;

  constructor(public payload: { authErrors: Errors }) { }
}

export class SetUpdateUserErrors implements Action {
  readonly type = AuthActionTypes.SetUpdateUserErrors;

  constructor(public payload: { authErrors: Errors }) { }
}

export class UpdateUserRequest implements Action {
  readonly type = AuthActionTypes.UpdateUserRequest;

  constructor(public payload: { user: User }) { }
}

export class UpdateUserSuccess implements Action {
  readonly type = AuthActionTypes.UpdateUserSuccess;

  constructor(public payload: { user: User }) { }
}

export class UpdateUserFail implements Action {
  readonly type = AuthActionTypes.UpdateUserFail;

  constructor(public payload: { errors: Errors }) { }
}

export type AuthActions =
  SetReturnUrl
  | ClearReturnStateFromRouteChange
  | AuthAttemptToGetUser
  | LoginSuccess
  | LoginFail
  | SettingsPageLogoutAction
  | LogoutConfirm
  | LogoutAction
  | LoginAttemptCopmlete
  | LoginPageAttemptLogin
  | LoggedLocalStorage
  | LoggedLocalStorageRequest
  | LoginPageClearAuthErrors
  | ClearAuthErrors
  | LoginPageSetAuthErrors
  | UpdateUserRequest
  | UpdateUserSuccess
  | UpdateUserFail
  | SetUpdateUserErrors;


