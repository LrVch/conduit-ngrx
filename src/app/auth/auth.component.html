<div class="auth-page page">
  <div class="container">
    <div class="row">

      <div class="col-md-6 offset-md-3 col-xs-12">
        <h1>{{ title }}</h1>
        <p>
          <a [routerLink]="['/login']"
             *ngIf="authType === 'register'">Have an account?</a>
          <a [routerLink]="['/register']"
             *ngIf="authType === 'login'">Need an account?</a>
        </p>

        <app-list-error [errors]="authErrors$ | async"></app-list-error>

        <form [formGroup]="authForm"
              (ngSubmit)="submitForm()">
          <fieldset [disabled]="isSubmitting$ | async">
            <fieldset class="">
              <mat-form-field *ngIf="authType === 'register'"
                              class="input-full-width">
                <input matInput
                       formControlName="username"
                       placeholder="Username"
                       type="text">
              </mat-form-field>
            </fieldset>
            <fieldset>
              <mat-form-field class="input-full-width">
                <input matInput
                       formControlName="email"
                       placeholder="Email"
                       type="text">
              </mat-form-field>
            </fieldset>
            <fieldset>
              <mat-form-field class="input-full-width">
                <input matInput
                       formControlName="password"
                       placeholder="Password"
                       type="password">
              </mat-form-field>
            </fieldset>
            <br>
            <button type="submit" [disabled]="!authForm.valid || (isSubmitting$ | async)" mat-flat-button color="primary"> {{ title }}</button>
            <br>
            <br>
          </fieldset>
        </form>
      </div>

    </div>
  </div>
</div>