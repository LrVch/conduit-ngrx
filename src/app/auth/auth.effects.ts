import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import {
  LoginSuccess,
  AuthActionTypes,
  LogoutAction,
  LoginFail,
  LoggedLocalStorage,
  LoginPageAttemptLogin,
  LoginAttemptCopmlete,
  LoginPageSetAuthErrors,
  UpdateUserRequest,
  UpdateUserSuccess,
  UpdateUserFail,
  SetUpdateUserErrors,
  LoggedLocalStorageRequest
} from './auth.actions';
import { tap, finalize, map, switchMap, catchError } from 'rxjs/operators';
import { Router } from '@angular/router';
import { Store, Action } from '@ngrx/store';
import { of, noop, defer, Observable } from 'rxjs';
import { UserService, User, Errors, ErrorsObj } from '../core';
import { AppState } from '../reducers';
import { JwtService } from '../core';
import { HideMainLoader } from '../layout/layout.actions';


@Injectable()
export class AuthEffects {

  @Effect({ dispatch: false })
  login$ = this.actions$.pipe(
    ofType<LoginSuccess>(AuthActionTypes.LoginSuccess),
    tap(action => {
      const { user } = action.payload;

      this.jwtService.saveToken(user.token);
      this.router.navigateByUrl('/');
    })
  );

  @Effect({ dispatch: false })
  logout$ = this.actions$.pipe(
    ofType<LogoutAction>(
      AuthActionTypes.LogoutAction,
    ),
    tap(() => {
      this.jwtService.destroyUseData();
      this.router.navigateByUrl('/login');
    })
  );

  @Effect({ dispatch: false })
  loginFail$ = this.actions$.pipe(
    ofType<LoginFail>(
      AuthActionTypes.LoginFail
    ),
    tap((action) => {
      const { authErrors } = action.payload;
      // console.error(authErrors);
      this.store.dispatch(new LoginPageSetAuthErrors({ authErrors }));
      this.store.dispatch(new HideMainLoader());
      this.jwtService.destroyUseData();
    })
  );

  @Effect()
  init$: Observable<Action> = defer((): Observable<LoggedLocalStorage | LoginFail> => {
    const token = this.jwtService.getToken();

    if (token) {
      this.store.dispatch(new LoggedLocalStorageRequest());
      return this.userService.getUser()
        .pipe(
          switchMap(responce => {
            const { user } = responce;
            return of(new LoggedLocalStorage({ user }));
          }),
          catchError(authErrors => {
            return of(new LoginFail({ authErrors }));
          })
        );
    } else {
      return of(new LoginFail({ authErrors: new ErrorsObj({ type: 'token', body: ['token not found'] }) }));
    }
  });

  @Effect()
  loginAttempt = this.actions$.pipe(
    ofType<LoginPageAttemptLogin>(AuthActionTypes.LoginPageAttemptLogin),
    switchMap(action => {
      return this.userService.attemptAuth(action.payload.authType, action.payload.credentials)
        .pipe(
          map(user => new LoginSuccess({ user })),
          catchError(authErrors => {
            return of(new LoginFail({ authErrors }));
          })
        );
    })
  );

  @Effect()
  updateInfo = this.actions$.pipe(
    ofType<UpdateUserRequest>(AuthActionTypes.UpdateUserRequest),
    switchMap(action => {
      return this.userService.update(action.payload.user)
        .pipe(
          map(user => new UpdateUserSuccess({ user })),
          catchError((errors: Errors) => {
            return of(new UpdateUserFail({ errors }));
          })
        );
    })
  );

  @Effect()
  updateInfoFail: Observable<Action> = this.actions$.pipe(
    ofType<UpdateUserFail>(AuthActionTypes.UpdateUserFail),
    map(action => new SetUpdateUserErrors({ authErrors: action.payload.errors }))
  );

  constructor(
    private actions$: Actions,
    private router: Router,
    private userService: UserService,
    private store: Store<AppState>,
    private jwtService: JwtService
  ) { }
}
