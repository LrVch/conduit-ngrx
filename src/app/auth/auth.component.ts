import { Component, OnInit, OnDestroy, } from '@angular/core';
import { ActivatedRoute, Router, RouterEvent, ResolveEnd, NavigationStart } from '@angular/router';

import { Errors } from '@app/core';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { LoginPageAttemptLogin, LoginPageClearAuthErrors, ClearReturnStateFromRouteChange } from './auth.actions';
import { Observable, Subject } from 'rxjs';
import { selectAuthLoading, selectAuthErrors } from './auth.selectors';
import { ShowMainLoader } from '@app/layout/layout.actions';
import { AuthPayload } from './auth-from/auth-from.component';
import { map, takeUntil } from 'rxjs/operators';

@Component({
  templateUrl: './auth.component.html'
})
export class AuthComponent implements OnInit, OnDestroy {
  authType$: Observable<string>;
  title$: Observable<string>;
  authErrors$: Observable<Errors>;
  isSubmitting$: Observable<boolean>;
  unsubscribe$ = new Subject<any>();

  BASE_LOGIN_URL = '/login';
  BASE_REGISTER_URL = '/register';

  constructor(
    private route: ActivatedRoute,
    private store: Store<AppState>,
    private router: Router
  ) { }

  ngOnInit() {
    this.authType$ = this.route.url.pipe(
      map(data => data[data.length - 1].path)
    );

    this.title$ = this.authType$.pipe(
      map(authType => authType === 'login' ? 'conduit.auth.sigin.title' : 'conduit.auth.sinup.title')
    );

    this.isSubmitting$ = this.store.pipe(
      select(selectAuthLoading)
    );

    this.authErrors$ = this.store.pipe(
      select(selectAuthErrors)
    );

    this.store.dispatch(new LoginPageClearAuthErrors());

    this.router.events.pipe(takeUntil(this.unsubscribe$)).subscribe((event: RouterEvent) => {
      if (event instanceof NavigationStart) {
        if (event.url.indexOf(this.BASE_LOGIN_URL) !== -1 || event.url.indexOf(this.BASE_REGISTER_URL) !== -1) {
        } else {
          this.store.dispatch(new ClearReturnStateFromRouteChange());
        }
      }
    });
  }

  ngOnDestroy() {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }

  submitForm(authPayload: AuthPayload): void {
    const { credentials, authType } = authPayload;

    this.store.dispatch(new LoginPageAttemptLogin({
      authType: authType,
      credentials
    }));

    this.store.dispatch(new ShowMainLoader());
  }
}
