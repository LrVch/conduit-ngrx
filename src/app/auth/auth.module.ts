import { NgModule } from '@angular/core';

import { AuthComponent } from './auth.component';
import { NoAuthGuard } from './no-auth-guard.service';
import { SharedModule } from '../shared';
import { AuthRoutingModule } from './auth-routing.module';
import { StoreModule } from '@ngrx/store';
import * as fromAuth from './auth.reducer';
import { EffectsModule } from '@ngrx/effects';
import { AuthEffects } from './auth.effects';

@NgModule({
  imports: [
    SharedModule,
    AuthRoutingModule,
    StoreModule.forFeature('auth', fromAuth.authReducer),
    EffectsModule.forFeature([AuthEffects])
  ],
  declarations: [
    AuthComponent
  ],
  providers: [
    NoAuthGuard
  ]
})
export class AuthModule {}
