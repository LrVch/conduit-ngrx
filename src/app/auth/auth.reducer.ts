import { User, Errors } from '../core';
import { AuthActions, AuthActionTypes } from './auth.actions';


export interface AuthState {
  loggedIn: boolean;
  loading: boolean;
  updatinInfo: boolean;
  authErrors: Errors;
  user: User;
  returnUrl: null | string;
}

export const initialAuthState: AuthState = {
  loggedIn: false,
  loading: false,
  authErrors: {
    errors: {}
  },
  user: null,
  updatinInfo: false,
  returnUrl: null
};

export function authReducer(
  state: AuthState = initialAuthState,
  action: AuthActions
): AuthState {
  switch (action.type) {
    case AuthActionTypes.LoginSuccess:
    case AuthActionTypes.LoggedLocalStorage:
      return {
        ...state,
        user: action.payload.user,
        loggedIn: true,
        loading: false
      };
    case AuthActionTypes.LoggedLocalStorageRequest:
      return {
        ...state,
        loggedIn: true,
      };
    case AuthActionTypes.LoginFail:
    case AuthActionTypes.LogoutConfirm:
    case AuthActionTypes.LogoutAction:
      return {
        ...state,
        loggedIn: false,
        user: null,
        loading: false,
        returnUrl: null
      };
    case AuthActionTypes.LoginPageAttemptLogin:
      return {
        ...state,
        loading: true,
        authErrors: {
          errors: {}
        }
      };
    case AuthActionTypes.LoginPageClearAuthErrors:
    case AuthActionTypes.ClearAuthErrors:
      return {
        ...state,
        authErrors: {
          errors: {}
        }
      };
    case AuthActionTypes.LoginPageSetAuthErrors:
    case AuthActionTypes.SetUpdateUserErrors:
      const { authErrors } = action.payload;
      return {
        ...state,
        authErrors
      };
    case AuthActionTypes.UpdateUserRequest:
      return {
        ...state,
        updatinInfo: true,
        authErrors: {
          errors: {}
        }
      };
    case AuthActionTypes.UpdateUserSuccess:
      return {
        ...state,
        updatinInfo: false,
        user: action.payload.user
      };
    case AuthActionTypes.UpdateUserFail:
      return {
        ...state,
        updatinInfo: false
      };
    case AuthActionTypes.SetReturnUrl:
      return {
        ...state,
        returnUrl: action.payload.returnUrl
      };
    case AuthActionTypes.ClearReturnState:
      return {
        ...state,
        returnUrl: null
      };
    default:
      return state;
  }
}

