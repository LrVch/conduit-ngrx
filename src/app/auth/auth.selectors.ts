import { createSelector, createFeatureSelector } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { AuthState } from './auth.reducer';

export const selectAuthState = createFeatureSelector<AppState, AuthState>('auth');

export const selectAuthLoggedIn = createSelector(
  selectAuthState,
  (authState: AuthState) => authState.loggedIn
);

export const selectAuthLoading = createSelector(
  selectAuthState,
  (authState: AuthState) => authState.loading
);

export const selectAuthErrors = createSelector(
  selectAuthState,
  (authState: AuthState) => authState.authErrors
);

export const selectUser = createSelector(
  selectAuthState,
  (authState: AuthState) => authState.user
);

export const selectUserUpdatingInfo = createSelector(
  selectAuthState,
  (authState: AuthState) => authState.updatinInfo
);

export const selectReturnUrl = createSelector(
  selectAuthState,
  (authState: AuthState) => authState.returnUrl
);
