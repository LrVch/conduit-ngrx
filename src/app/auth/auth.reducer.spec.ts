import * as fromAuth from './auth.reducer';
import {
  LoginSuccess,
  LoggedLocalStorage,
  LoggedLocalStorageRequest,
  LoginFail,
  LogoutConfirm,
  LogoutAction,
  LoginPageAttemptLogin,
  LoginPageClearAuthErrors,
  ClearAuthErrors,
  LoginPageSetAuthErrors,
  SetUpdateUserErrors,
  UpdateUserRequest,
  UpdateUserSuccess,
  UpdateUserFail,
  SetReturnUrl,
  ClearReturnStateFromRouteChange
} from './auth.actions';
import { getUser, getAuthErrors, getCredentials } from '@app/lib/testing';

describe('Auth Reducer', () => {
  const user = getUser();
  const authErrors = getAuthErrors();
  const errors = getAuthErrors();
  const initialAuthState: fromAuth.AuthState = {
    loggedIn: false,
    loading: false,
    authErrors: {
      errors: {}
    },
    user: null,
    updatinInfo: false,
    returnUrl: null
  };
  const credentials = getCredentials();

  describe('State changes', () => {
    it('should have an initial state', () => {
      const state = fromAuth.authReducer(initialAuthState, {
        type: '@@init'
      } as any);

      expect(state).toBe(initialAuthState);
    });

    it('should load user on success', () => {
      const init = { type: '@@init' } as any;
      const loginSuccess = new LoginSuccess({ user });

      const state = [init, loginSuccess].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });

    it('should load user on localstorage logged success', () => {
      const init = { type: '@@init' } as any;
      const loggedLocalStorage = new LoggedLocalStorage({ user });

      const state = [init, loggedLocalStorage].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });

    it('should set "loggedIn" to true on "LoggedLocalStorageRequest"', () => {
      const init = { type: '@@init' } as any;
      const loggedLocalStorageRequest = new LoggedLocalStorageRequest();

      const state = [init, loggedLocalStorageRequest].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });

    it('should set "loggedIn" to true on "LoggedLocalStorageRequest"', () => {
      const init = { type: '@@init' } as any;
      const loggedLocalStorageRequest = new LoggedLocalStorageRequest();

      const state = [init, loggedLocalStorageRequest].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });

    it('should logout user on "LoginFail" or "LogoutConfirm" or "LogoutAction"', () => {
      const init = { type: '@@init' } as any;
      const loginFail = new LoginFail({ authErrors });
      const logoutConfirm = new LogoutConfirm();
      const logoutAction = new LogoutAction();

      const state1 = [init, loginFail].reduce(
        fromAuth.authReducer,
        initialAuthState
      );
      const state2 = [init, logoutConfirm].reduce(
        fromAuth.authReducer,
        initialAuthState
      );
      const state3 = [init, logoutAction].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state1).toMatchSnapshot();
      expect(state2).toMatchSnapshot();
      expect(state3).toMatchSnapshot();
    });

    it('should reset auth error and set "loading" to true on "LoginPageAttemptLogin"', () => {
      const init = { type: '@@init' } as any;
      const loginPageAttemptLogin = new LoginPageAttemptLogin({
        authType: '',
        credentials
      });

      const state = [init, loginPageAttemptLogin].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });

    it('should clear auth errors', () => {
      const init = { type: '@@init' } as any;
      const loginPageClearAuthErrors = new LoginPageClearAuthErrors();
      const clearAuthErrors = new ClearAuthErrors();

      const state1 = [init, loginPageClearAuthErrors].reduce(
        fromAuth.authReducer,
        initialAuthState
      );
      const state2 = [init, clearAuthErrors].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state1).toMatchSnapshot();
      expect(state2).toMatchSnapshot();
    });

    it('should set auth errors', () => {
      const init = { type: '@@init' } as any;
      const loginPageClearAuthErrors = new LoginPageSetAuthErrors({
        authErrors
      });
      const setUpdateUserErrors = new SetUpdateUserErrors({ authErrors });

      const state1 = [init, loginPageClearAuthErrors].reduce(
        fromAuth.authReducer,
        initialAuthState
      );
      const state2 = [init, setUpdateUserErrors].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state1).toMatchSnapshot();
      expect(state2).toMatchSnapshot();
    });

    it('should set "updatinInfo" to true on "UpdateUserRequest"', () => {
      const init = { type: '@@init' } as any;
      const updateUserRequest = new UpdateUserRequest({ user });

      const state1 = [init, updateUserRequest].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state1).toMatchSnapshot();
    });

    it('should set "updatinInfo" to false on "UpdateUserFail"', () => {
      const init = { type: '@@init' } as any;
      const updateUserFail = new UpdateUserFail({ errors });

      const state1 = [init, updateUserFail].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state1).toMatchSnapshot();
    });

    it('should set user on "UpdateUserSuccess"', () => {
      const init = { type: '@@init' } as any;
      const updateUserSuccess = new UpdateUserSuccess({ user });

      const state1 = [init, updateUserSuccess].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state1).toMatchSnapshot();
    });

    it('should set returnUrl on "SetReturnUrl"', () => {
      const init = { type: '@@init' } as any;
      const setReturnUrl = new SetReturnUrl({ returnUrl: 'returnUrl' });

      const state = [init, setReturnUrl].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });

    it('should clear returnUrl on "ClearReturnStateFromRouteChange"', () => {
      const init = { type: '@@init' } as any;
      const clearReturnUrl = new ClearReturnStateFromRouteChange();

      const state = [init, clearReturnUrl].reduce(
        fromAuth.authReducer,
        initialAuthState
      );

      expect(state).toMatchSnapshot();
    });
  });
});
