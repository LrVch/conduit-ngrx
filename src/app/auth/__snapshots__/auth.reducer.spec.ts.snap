// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Reducer State changes should ckear returnUrl on "ClearReturnUrl" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should clear auth errors 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should clear auth errors 2`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should clear returnUrl on "ClearReturnStateFromRouteChange" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should load user on localstorage logged success 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": true,
  "returnUrl": null,
  "updatinInfo": false,
  "user": Object {
    "bio": "bil",
    "email": "email",
    "image": "image",
    "token": "token",
    "username": "usernam",
  },
}
`;

exports[`Auth Reducer State changes should load user on success 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": true,
  "returnUrl": null,
  "updatinInfo": false,
  "user": Object {
    "bio": "bil",
    "email": "email",
    "image": "image",
    "token": "token",
    "username": "usernam",
  },
}
`;

exports[`Auth Reducer State changes should logout user on "LoginFail" or "LogoutConfirm" or "LogoutAction" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should logout user on "LoginFail" or "LogoutConfirm" or "LogoutAction" 2`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should logout user on "LoginFail" or "LogoutConfirm" or "LogoutAction" 3`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should reset auth error and set "loading" to true on "LoginPageAttemptLogin" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": true,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set "loggedIn" to true on "LoggedLocalStorageRequest" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": true,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set "loggedIn" to true on "LoggedLocalStorageRequest" 2`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": true,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set "updatinInfo" to false on "UpdateUserFail" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set "updatinInfo" to true on "UpdateUserRequest" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": true,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set auth errors 1`] = `
Object {
  "authErrors": ErrorsObj {
    "errors": Object {
      "error": Array [
        "some error",
      ],
    },
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set auth errors 2`] = `
Object {
  "authErrors": ErrorsObj {
    "errors": Object {
      "error": Array [
        "some error",
      ],
    },
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set returnUrl on "SetReturnUrl" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": "returnUrl",
  "updatinInfo": false,
  "user": null,
}
`;

exports[`Auth Reducer State changes should set user on "UpdateUserSuccess" 1`] = `
Object {
  "authErrors": Object {
    "errors": Object {},
  },
  "loading": false,
  "loggedIn": false,
  "returnUrl": null,
  "updatinInfo": false,
  "user": Object {
    "bio": "bil",
    "email": "email",
    "image": "image",
    "token": "token",
    "username": "usernam",
  },
}
`;
